Dockerfile

```diff
@@ -1,71 +1,89 @@
-
 RUN bundle config set without "development test"
-ARG BUNDLE_FLAGS="--jobs=4 --no-binstubs --no-cache"
-RUN bundle install ${BUNDLE_FLAGS}
+RUN bundle install --jobs=4 --no-binstubs --no-cache
 
-# Install node packages defined in package.json, including webpack
 COPY package.json yarn.lock /app/
 RUN yarn install --frozen-lockfile
 
-# Copy all files to /app (except what is defined in .dockerignore)
 COPY . /app/
 
-# Compile assets and run webpack. We set a dummy secret key.
-RUN RAILS_ENV=production bundle exec rails SECRET_KEY_BASE=pickasecuretoken assets:precompile
+RUN bundle exec rake assets:precompile
 
-# Cleanup to save space in the production image
-RUN rm -rf node_modules log tmp && \
-      rm -rf /usr/local/bundle/cache && \
-      rm -rf .env && \
-      find /usr/local/bundle/gems -name "*.c" -delete && \
-      find /usr/local/bundle/gems -name "*.h" -delete && \
-      find /usr/local/bundle/gems -name "*.o" -delete && \
-      find /usr/local/bundle/gems -name "*.html" -delete
+ENV PORT=8080
+EXPOSE ${PORT}
 
-# Build runtime image
-FROM ruby:3.1.4-alpine as production
+RUN adduser --system --no-create-home nonroot
 
-# The application runs from /app
-WORKDIR /app
+# We expect the rake assets:precompile command to create these directories, but mkdir -p will create them if they don't already exist
+RUN mkdir -p tmp log
+RUN chown -R nonroot tmp log
+RUN chown nonroot db/schema.rb
 
-# Add postgres driver library
-# Add the timezone as it's not configured by default in Alpine
-RUN apk add --update --no-cache libpq tzdata && \
-  cp /usr/share/zoneinfo/Europe/London /etc/localtime && \
-  echo "Europe/London" > /etc/timezone
+FROM base as development
 
-# Copy files generated in the builder image
-COPY --from=builder /app /app
-COPY --from=builder /usr/local/bundle/ /usr/local/bundle/
+RUN bundle config set without ""
+RUN bundle install --jobs=4 --no-binstubs --no-cache
 
-ARG GIT_COMMIT_SHA="UNKNOWN"
-ENV GIT_COMMIT_SHA=${GIT_COMMIT_SHA}
-RUN echo ${GIT_COMMIT_SHA} > ./GIT_COMMIT_SHA
+# Install gecko driver for Capybara tests
+RUN apk add firefox
+RUN wget https://github.com/mozilla/geckodriver/releases/download/v0.31.0/geckodriver-v0.31.0-linux64.tar.gz \
+    && tar -xvzf geckodriver-v0.31.0-linux64.tar.gz \
+    && rm geckodriver-v0.31.0-linux64.tar.gz \
+    && chmod +x geckodriver \
+    && mv geckodriver /usr/local/bin/
 
-ENV PORT=8080
+USER nonroot
 
-EXPOSE ${PORT}
+CMD bundle exec rails s -e ${RAILS_ENV} -p ${PORT} --binding=0.0.0.0
+
+FROM base as staging
+
+RUN bundle config set without "development"
+RUN bundle install --jobs=4 --no-binstubs --no-cache
+
+# Cleanup to reduce image size
+RUN rm -rf node_modules && \
+    rm -rf /usr/local/bundle/cache && \
+    find /usr/local/bundle/gems -name "*.c" -delete && \
+    find /usr/local/bundle/gems -name "*.h" -delete && \
+    find /usr/local/bundle/gems -name "*.o" -delete && \
+    find /usr/local/bundle/gems -name "*.html" -delete && \
+    rm -rf .env  && \
```

Comment:
Shouldn't `.env` not even be copied into the image in the first place? And it still exists in the previous layer, so you're leaking secrets (though only dev ones) in the image.

Reply:
In response to Hugh's suggestion "Even if you're not yet pushing the image to ECR, you should still get GitHub actions to check the image builds." - we will do this later in the project when updating/building the deployment pipelines. That's out of scope for this ticket.